### ISSUE TITLE
[Feature] [RLlib] Pytorch JIT as framework

### ISSUE THREAD

**Issue by smorad** [ray-project/ray#22166]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

According to Sven, the torch framework is roughly 50% slower than the tf framework. We should look into `torch.jit` to see if we can get any speedups.

We have a couple different options:
`torch.jit.trace`: This traces a model using dummy tensors, but it can miss unused portions of the code path (i.e. conditionals). I believe the surface for errors here is large, and believe we should stay away from it.

`torch.jit.script`: This does static analysis on the python source and compiles it into `TorchScript`. This is likely the way forward, but to do it correctly, it requires all class members to be `TorchScript` compatible which is not feasible given the current `TorchModelV2` (e.g. any code utilizing `numpy` arrays is not compilable).

`torch.jit.script` submodules: Rather than JIT the entire `TorchModelV2`, we can jit all `torch.nn.Module` children of `TorchModelV2`. This allows us to use the JIT without too many changes. The downside is we lose a few optimizations between distinct networks (e.g. with MLP -> Value net, we cannot fuse the last MLP operation with the first value netw operation). I think this is the best option, because the number of optimizations we lose is very small.

`torch.jit.script` submodules with `TensorRT`: Like above, but utilize `TensorRT` from NVidia to further lower TorchScript into native TensorRT CUDA code. This would presumably only work on NVidia GPUs with Tensor cores, and would require using reduced-precision (`fp16`) floats, but could provide significant speedups for the env sampling workers: https://developer.nvidia.com/blog/accelerating-inference-up-to-6x-faster-in-pytorch-with-torch-tensorrt/

`torch.jit.script` submodules with `CudaGraphs`: Like above, but utilize `torch` to construct the `graph` on a CUDA device by capturing the CUDA stream. Rather than call into the model normally, we can load new arguments into GPU buffers and replay the stream with the new arguments: https://pytorch.org/docs/master/notes/cuda.html#cuda-graphs

`torch.jit.script` submodules with `torch.AMP`: Like above, but automatically cast all tensors to `fp16`. This utilizes CUDA tensor cores for a 4x speedup (and some CPU equivalent for a less-modest speedup), but requires scaling gradients using `GradScaler` https://pytorch.org/docs/stable/amp.html.

### Use case

`torch.jit` the default models.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

**Comment by smorad** [comment]
Preliminary results, using `torch.jit.script` submodules approach:
<img width="376" alt="Screen Shot 2022-02-07 at 11 17 10 AM" src="https://user-images.githubusercontent.com/1743889/152778348-502d734e-8c96-451a-9d88-6365885474dc.png">
Orange is JIT blue is without JIT -- the task is stateless cartpole using the default LSTM model. It seems like the JIT compilation cost dominates early, making it slower than without jit. At around 400k timesteps, it seems like they are roughly even (maybe without JIT is a little faster). It's quite surprising that JITting the code can make it slower, but I've done some googling and it seems like other people see this at times as well.

Note that these are CPU times as my laptop doesn't have a GPU. It's possible these results change significantly when using a GPU, as multiple operations are fused into a single CUDA kernel.

**Comment by smorad** [comment]
I have done testing of AMP and JIT on a Tesla T4 GPU, using an AWS `g4dx.2xlarge`, the results are available at https://tensorboard.dev/experiment/OYJwoaI5Qji63JTWENgleA/ using the following config
```
config = {
    "env": "CartPole-v0",
    "num_workers": 2,
    "num_envs_per_worker": 8,
    "framework": "torch",
    "num_gpus": 0.5,
    "num_gpus_per_worker": 0.25,
    "model": {
        "fcnet_hiddens": [512, 512, 512, 512],
    },
}
```

The small model uses the default `[256, 256]` hidden layers. The large model uses `[512, 512, 512, 512]` hidden layers.

The summary is that both JIT and AMP are quite disappointing. With AMP, we actually see a slowdown. The speedup we get from `float16` is outweighed by the cost of dynamic casting. With JIT, performance is roughly equal for nearly all models. There is an exception -- JIT IMPALA big performs slightly better than default IMPALA big. We see a 6% speedup in the total experiment time.

It's important to note that during testing, GPU utilization did not exceed 55%, even with the sample collectors on the GPU.

My conclusion is that while AMP/JIT do produce significant speedups, these are usually in big vision models (e.g. ResNet-101 has 101 layers). For the type of models we use in RL (2-5 layers), I suspect the time we actually spend in the model forward/backward passes is negligible compared to all the RLlib machinery surrounding sample collection and training. To gain significant performance benefits from either AMP or JIT, we would need to do at least one of the following:

1. Increase batch size significantly and do away with minibatching
2. Use deep (>10 layers) and wide (>512) models

**Comment by gjoliver** [comment]
if CPU is the bottleneck, we can try running the 2 workers on 2 different nodes?
we also track various timing stats. for example, worker stats:
https://github.com/ray-project/ray/blob/c3e04ab275b8367b87eb8ceb60a8084ed647b468/rllib/evaluation/sampler.py#L87-L91
and there are similar fields for learning as well.
will be interesting to actually see some of the numbers.

**Comment by smorad** [comment]
@gjoliver all these numbers are tracked in the tensorboard link in my last post. I used a `g4dn.2xlarge` which has 8 CPUs -- that should be enough for 2 workers + trainer, right? I suspect we'll see even slower performance moving workers off of the trainer host, because then `SampleBatches` need to be serialized and sent over the network.

**Comment by gjoliver** [comment]
ah right. looking at everything under /sampler_perf/, mean_inference_ms is still the dominating component, although it only takes about 1.7ms, and no big difference between jit and non-jit version.
try a really big network? the first step probably should be to observe some performance gain for the inference step alone.

**Comment by smorad** [comment]
@gjoliver I'm wondering if huge models are a valid use case. In research, most practitioners use somewhere between 2-4 hidden layers. Are there any commercial users of RLlib that are concerned with truly deep models? If so, it makes sense to pursue this. If not, then I think efforts could be better spent elsewhere.

**Comment by gjoliver** [comment]
totally agree. I only know 1 user who is using a network maybe 15 to 20 layers deep. mostly pretty small.
this is mostly for satisfying our own curiosity, like is there any artificial scenario we can create where this shows any kind of visible benefit?
otherwise, I wonder maybe we did something stupid before that breaks all these nice improvements ...

but obviously not super high priority. we can pick these up sometime in the future.

**Comment by gjoliver** [comment]
let's keep this open.

**Comment by smorad** [comment]
It seems like control-flow operators will [prevent fusion](https://github.com/pytorch/pytorch/issues/75464). We can try using the `torch.jit.fuser` to force fusion of our models.

Another big issue is that the kernel must be recompiled when the batch size changes, so we must ensure this remains constant.

**Comment by smorad** [comment]
One more approach would be to speedup inference by converting the `jit`ted model to a TensorRT graph: https://github.com/pytorch/TensorRT

However, this would only speed up inference on GPU, which is not something we currently do in RLlib. Only once we can effectively batch rollout worker environment steps (e.g. since worker with 256 envs, stepped in parallel in a single batch) would we have a reason to use a GPU for inference.

**Comment by smorad** [comment]
With the new `torch.compile` present in the upcoming torch 2.0,  I think we can close this.
